src/NHibernate.Test/QueryTest/MultiCriteriaFixture.cs

```diff
@@ -487,5 +487,38 @@ public void CanGetResultInAGenericList()
 				Assert.That(results[1], Is.InstanceOf<List<int>>());
 			}
 		}
+
+		[Test]
+		public void UsingManyParametersAndQueries_DoesNotCauseParameterNameCollisions()
+		{
+			//GH-1357
+			using (var s = OpenSession())
+			{
+				var item = new Item {Id = 15};
+				s.Save(item);
+				s.Flush();
+			}
+
+			using (var s = OpenSession())
+			{
+				var multi = s.CreateMultiCriteria();
+
+				for (var i = 0; i < 12; i++)
+				{
+					var criteria = s.CreateCriteria(typeof(Item));
+					for (var j = 0; j < 12; j++)
+					{
+						criteria = criteria.Add(Restrictions.Gt("id", j));
+					}
+					multi.Add(criteria);
+				}
+				//Paramater combining is only used for cacheable queries
+				multi.SetCacheable(true);
```

Comment:
Small typo (Paramater instead of Parameter) and yet another somewhat undue implementation discrepancy between criteria and query. I think the criteria implementation is the right one, the query one creates those parameters which are only used in the query cache key uselessly when the query is not cacheable.

So if we realign/mutualize implementation later, the multi-query test may no more test anything. It would be better to enable caching on it too (by enabling caching on all its queries, with same region).

Reply:
Good idea. I'll fix the typo, set cacheable and rebase.